File: app/src/tools/edit.ts
import { resolve } from "node:path";
import type { AgentTool } from "../runtime/agent";
import { Type } from "../runtime/agent";
import { defaultFileOperations, type FileOperations } from "./file-operations";

// Legacy single-edit shape the model sometimes sends instead of the canonical {edits:[...]} shape.
type LegacyEditParams = {
	path: string;
	oldText: string;
	newText: string;
	edits?: never;
};
export type ExactEdit = { oldText: string; newText: string };

type CanonicalEditParams = {
	path: string;
	edits: ExactEdit[];
};
type EditParams = CanonicalEditParams | LegacyEditParams;

export const exactEditsSchema = Type.Array(
	Type.Object({
		oldText: Type.String({
			description:
				"Exact text for one targeted replacement. Must be unique in the file and must not overlap with other edits in the same call.",
		}),
		newText: Type.String({
			description: "Replacement text for this targeted edit.",
		}),
	}),
	{
		description:
			"One or more targeted replacements. Each edit is matched against the original file, not incrementally. Merge nearby changes into one edit instead of overlapping edits.",
		minItems: 1,
	},
);

const editSchema = Type.Object({
	path: Type.String({
		description: "Path to the file to edit (relative or absolute)",
	}),
	edits: exactEditsSchema,
});

type EditRange = ExactEdit & { index: number; start: number; end: number };

function findMatchOffsets(content: string, search: string): number[] {
	if (search.length === 0) return [];
	const offsets: number[] = [];
	let from = 0;
	while (from <= content.length - search.length) {
		const index = content.indexOf(search, from);
		if (index < 0) break;
		offsets.push(index);
		from = index + 1;
	}
	return offsets;
}

export function applyExactEdits(
	original: string,
	edits: ExactEdit[],
): { content: string; errors: string[] } {
	const errors: string[] = [];
	const ranges: EditRange[] = [];
	for (let i = 0; i < edits.length; i++) {
		const edit = edits[i];
		if (edit.oldText.length === 0) {
			errors.push(`edits[${i}]: oldText must not be empty`);
			continue;
		}
		const offsets = findMatchOffsets(original, edit.oldText);
		if (offsets.length === 0) {
			errors.push(`edits[${i}]: oldText not found in file`);
		} else if (offsets.length > 1) {
			errors.push(
				`edits[${i}]: oldText matches ${offsets.length} locations — must be unique`,
			);
		} else {
			const start = offsets[0];
			ranges.push({
				...edit,
				index: i,
				start,
				end: start + edit.oldText.length,
			});
		}
	}

	for (let i = 0; i < ranges.length; i++) {
		for (let j = i + 1; j < ranges.length; j++) {
			const left = ranges[i];
			const right = ranges[j];
			if (left.start < right.end && right.start < left.end) {
				errors.push(
					`edits[${left.index}] and edits[${right.index}] overlap — merge them into one edit`,
				);
			}
		}
	}

	if (errors.length > 0) return { content: original, errors };
	let content = original;
	for (const range of ranges.sort((a, b) => b.start - a.start)) {
		content =
			content.slice(0, range.start) + range.newText + content.slice(range.end);
	}
	return { content, errors };
}

export function createEditTool(
	cwd: string,
	files: FileOperations = defaultFileOperations,
): AgentTool<typeof editSchema> {
	return {
		name: "edit",
		label: "Edit",
		description:
			"Edit a file using exact text replacements. Each edit's oldText must match exactly and be unique. Multiple edits are applied to the original file simultaneously — do not include overlapping edits.",
		parameters: editSchema,

		async execute(_id, rawParams, _signal) {
			// Accept legacy single-edit shape {path, oldText, newText} in addition to
			// the canonical {path, edits:[...]} shape, since models sometimes use both.
			const raw = rawParams as unknown as EditParams;
			const legacy = raw as LegacyEditParams;
			const params: CanonicalEditParams =
				typeof legacy.oldText === "string" && !Array.isArray(raw.edits)
					? {
							path: raw.path,
							edits: [{ oldText: legacy.oldText, newText: legacy.newText }],
						}
					: (raw as CanonicalEditParams);
			const abs = resolve(cwd, params.path);
			const errors: string[] = [];
			try {
				await files.mutate(abs, (original) => {
					const result = applyExactEdits(original, params.edits);
					errors.push(...result.errors);
					if (errors.length > 0) throw new Error("Invalid edits");
					return result.content;
				});

				return {
					content: [
						{
							type: "text",
							text: `Applied ${params.edits.length} edit${params.edits.length === 1 ? "" : "s"} to ${params.path}`,
						},
					],
					details: { path: abs, applied: params.edits.length, errors: [] },
				};
			} catch (err) {
				if (errors.length > 0) {
					return {
						content: [{ type: "text", text: `Error:\n${errors.join("\n")}` }],
						details: { path: abs, applied: 0, errors },
					};
				}
				const msg = err instanceof Error ? err.message : String(err);
				return {
					content: [{ type: "text", text: `Error: ${msg}` }],
					details: { path: params.path, applied: 0, errors: [msg] },
				};
			}
		},
	};
}

File: internal/codingtools/edit.go
package codingtools

import (
	"context"
	"fmt"
	"sort"
	"strings"

	"github.com/akonwi/kit/internal/droids"
)

const maxEditFileBytes = 32 << 20

type exactEdit struct {
	OldText string
	NewText string
}

type editInput struct {
	OldText *string `json:"oldText"`
	NewText *string `json:"newText"`
}

type editArgs struct {
	Path    string      `json:"path"`
	Edits   []editInput `json:"edits,omitempty"`
	OldText *string     `json:"oldText,omitempty"`
	NewText *string     `json:"newText,omitempty"`
}

type editDetails struct {
	Path    string   `json:"path"`
	Applied int      `json:"applied"`
	Errors  []string `json:"errors"`
}

type editRange struct {
	exactEdit
	index int
	start int
	end   int
}

func newEditTool(cwd string) droids.Tool[editArgs] {
	return droids.Tool[editArgs]{
		Name:        EditToolName,
		Description: "Edit a file using exact text replacements. Each edit's oldText must match exactly and be unique. Multiple edits are applied to the original file simultaneously — do not include overlapping edits.",
		Parameters: objectSchema(map[string]any{
			"path": stringSchema("Path to the file to edit (relative or absolute)"),
			"edits": map[string]any{
				"type":        "array",
				"description": "One or more targeted replacements. Each edit is matched against the original file, not incrementally. Merge nearby changes into one edit instead of overlapping edits.",
				"minItems":    1,
				"items": objectSchema(map[string]any{
					"oldText": stringSchema("Exact text for one targeted replacement. Must be unique in the file and must not overlap with other edits in the same call."),
					"newText": stringSchema("Replacement text for this targeted edit."),
				}, "oldText", "newText"),
			},
		}, "path", "edits"),
		Mode: droids.ModeSequential,
		Execute: func(ctx context.Context, args editArgs) (droids.ToolResult, error) {
			if err := ctx.Err(); err != nil {
				return droids.ToolResult{}, err
			}
			path, err := resolvePath(cwd, args.Path)
			if err != nil {
				return errorResult(err, editDetails{Path: args.Path}), nil
			}
			edits, err := normalizedEdits(args)
			if err != nil {
				return editFailure(path, []string{err.Error()}), nil
			}
			var applied int
			var editErrors []string
			err = withMutationLock(ctx, path, func() error {
				if err := ctx.Err(); err != nil {
					return err
				}
				original, version, err := readRegularFileVersion(ctx, path, maxEditFileBytes)
				if err != nil {
					return err
				}
				updated, errors := applyExactEdits(string(original), edits)
				if len(errors) > 0 {
					editErrors = errors
					return nil
				}
				if updated != string(original) {
					if err := replaceRegularFile(ctx, path, []byte(updated), 0o666, &version); err != nil {
						return err
					}
				}
				applied = len(edits)
				return nil
			})
			if err != nil {
				if ctx.Err() != nil {
					return droids.ToolResult{}, ctx.Err()
				}
				return errorResult(err, editDetails{Path: path, Errors: []string{err.Error()}}), nil
			}
			if len(editErrors) > 0 {
				return editFailure(path, editErrors), nil
			}
			return textResult(
				fmt.Sprintf("Applied %d edit%s to %s", applied, plural(applied), args.Path),
				editDetails{Path: path, Applied: applied, Errors: []string{}},
			), nil
		},
	}
}

func normalizedEdits(args editArgs) ([]exactEdit, error) {
	if len(args.Edits) > 0 {
		edits := make([]exactEdit, 0, len(args.Edits))
		for index, edit := range args.Edits {
			if edit.OldText == nil {
				return nil, fmt.Errorf("edits[%d]: oldText is required", index)
			}
			if edit.NewText == nil {
				return nil, fmt.Errorf("edits[%d]: newText is required", index)
			}
			edits = append(edits, exactEdit{OldText: *edit.OldText, NewText: *edit.NewText})
		}
		return edits, nil
	}
	if args.OldText != nil && args.NewText != nil {
		return []exactEdit{{OldText: *args.OldText, NewText: *args.NewText}}, nil
	}
	return nil, fmt.Errorf("edits must contain at least one replacement")
}

func applyExactEdits(original string, edits []exactEdit) (string, []string) {
	errors := make([]string, 0)
	ranges := make([]editRange, 0, len(edits))
	for index, edit := range edits {
		if edit.OldText == "" {
			errors = append(errors, fmt.Sprintf("edits[%d]: oldText must not be empty", index))
			continue
		}
		offsets := matchOffsets(original, edit.OldText)
		switch len(offsets) {
		case 0:
			errors = append(errors, fmt.Sprintf("edits[%d]: oldText not found in file", index))
		case 1:
			ranges = append(ranges, editRange{exactEdit: edit, index: index, start: offsets[0], end: offsets[0] + len(edit.OldText)})
		default:
			errors = append(errors, fmt.Sprintf("edits[%d]: oldText matches %d locations — must be unique", index, len(offsets)))
		}
	}
	for left := 0; left < len(ranges); left++ {
		for right := left + 1; right < len(ranges); right++ {
			if ranges[left].start < ranges[right].end && ranges[right].start < ranges[left].end {
				errors = append(errors, fmt.Sprintf("edits[%d] and edits[%d] overlap — merge them into one edit", ranges[left].index, ranges[right].index))
			}
		}
	}
	if len(errors) > 0 {
		return original, errors
	}
	sort.Slice(ranges, func(left, right int) bool { return ranges[left].start > ranges[right].start })
	updated := original
	for _, replacement := range ranges {
		updated = updated[:replacement.start] + replacement.NewText + updated[replacement.end:]
	}
	return updated, nil
}

func matchOffsets(content, search string) []int {
	var offsets []int
	for from := 0; from <= len(content)-len(search); {
		index := strings.Index(content[from:], search)
		if index < 0 {
			break
		}
		index += from
		offsets = append(offsets, index)
		from = index + 1
	}
	return offsets
}

func editFailure(path string, errors []string) droids.ToolResult {
	return droids.ToolResult{
		Content: []droids.Content{droids.TextContent{Text: "Error:\n" + strings.Join(errors, "\n")}},
		Details: editDetails{Path: path, Errors: errors},
		IsError: true,
	}
}

func plural(count int) string {
	if count == 1 {
		return ""
	}
	return "s"
}
